services/identity/SitesService.ts

```diff
@@ -1,43 +1,104 @@
+
+import { BadRequestError } from "@errors/BadRequestError"
 
-import TokenStore from "./TokenStore"
+import { Site } from "@database/models"
 
-interface SitesServiceProps {
+export interface SitesServiceProps {
   repository: ModelStatic<Site>
-  tokenStore: TokenStore
+}
+
+export interface CreateSiteProps {
+  repositoryName: string
+  createdBy?: string
+  agency?: string
+  siteName?: string
+  contact?: string
+  repositoryUrl?: string
+  hostingId?: string
+  stagingUrl?: string
+  productionUrl?: string
+  liveDomain?: string
+  redirectFrom?: string[]
+  uptimeId?: string
+  uptimeUrl?: string
+  launchedAt?: string
+  launchedBy?: string
+}
+
+export interface UpdateSiteProps {
+  repositoryName?: string // Don't allow null
+  createdBy?: string | null
+  agency?: string | null
+  siteName?: string | null
+  contact?: string | null
+  repositoryUrl?: string | null
+  hostingId?: string | null
+  stagingUrl?: string | null
+  productionUrl?: string | null
+  liveDomain?: string | null
+  redirectFrom?: string[] | null
+  uptimeId?: string | null
+  uptimeUrl?: string | null
+  launchedAt?: string | null
+  launchedBy?: string | null
 }
 
 class SitesService {
   // NOTE: Explicitly specifying using keyed properties to ensure
   // that the types are synced.
   private readonly repository: SitesServiceProps["repository"]
 
-  private readonly tokenStore: SitesServiceProps["tokenStore"]
-
-  constructor({ repository, tokenStore }: SitesServiceProps) {
+  constructor({ repository }: SitesServiceProps) {
     this.repository = repository
-    this.tokenStore = tokenStore
   }
 
-  async getBySiteName(siteName: string): Promise<Site | null> {
-    const site = await this.repository.findOne({
-      where: { name: siteName },
-    })
+  async createSite(props: CreateSiteProps): Promise<Site> {
+    try {
+      return await Site.create(props as never) // Work-around for sequelize's useless type checking
+    } catch (err) {
+      if (err instanceof ValidationError) {
+        const items = err.errors.map((item) => item.message).join(", ")
+        throw new BadRequestError(`${err}: ${items}`)
+      } else {
+        throw err
+      }
+    }
+  }
 
-    return site
+  async getByRepositoryName(repositoryName: string): Promise<Site | null> {
+    return this.repository.findOne({
+      where: { repositoryName },
+    })
   }
 
-  async getSiteAccessToken(siteName: string) {
-    const site = await this.getBySiteName(siteName)
+  async updateSite(
+    repositoryName: string,
+    props: UpdateSiteProps
+  ): Promise<Site | null> {
+    try {
+      const [affectedRows, data] = await this.repository.update(props, {
+        where: { repositoryName },
+        returning: true,
+      })
 
-    if (!site) {
-      return null
+      if (affectedRows === 0) {
+        return null
+      }
+      if (affectedRows === 1) {
+        return data[0]
+      }
+      throw new Error(`Updated ${affectedRows} sites (expected 1).`)
```

Comment:
not sure this works as expected - if we get >1 affected row, will the update go through? ie, data is still modified in the database despite throwing an error